### PR TITLE
feat: remote MCP deployment with per-user OAuth 2.1 PKCE

### LINKED ISSUE #1: Allow to deploy as a Remote MCP
Remote MCP are much better to use.
This MCP should support to be deployed as a remote MCP server on any HTTP provider

Compatibly with local deployment required 
Cli arguments must be available through:
- either static startup config (env vars)
- or dymanic config (use cases and activation to be determined)

Documentation must be adapted to separate personas:
- the one which is looking for a local MCP server
- the one which is looking for a Zendesk MCP server which could be deployed on a server to expose a private remote MCP server on their Zendesk account

This feature is not about deploying a live remote MCP server

Optional : must be compatible with glama requirements which allows to deploy existing MCP server and expose them as live remote MCP server

### PR DESCRIPTION
**Pull request by dlecan**
Closes #1.

> **HTTP transport ships as experimental.** Stdio is the supported path and is the one the local gate exercises end-to-end. The HTTP transport works in unit tests, in the smoke test, and against a local `curl`, but its end-to-end behavior against a real Zendesk tenant from every supported MCP client has not yet been validated. Until that is done, expect possible rough edges around OAuth discovery behind reverse proxies, CORS with browser-based clients, and the 401 / refresh flow. The same notice is on the corresponding README section so operators see it before deploying.

## Summary

Adds an HTTP transport so the same binary can run as a local stdio MCP server (laptop, Claude Desktop / Code / VS Code) or as a private remote MCP server (Claude.ai web, ChatGPT, etc.). The per-user OAuth 2.1 PKCE security model is preserved in both modes; API token auth is refused in HTTP because a shared static credential reachable over the network defeats the per-user model.

## Type of change
- [ ] Bug fix
- [x] New feature
- [ ] Refactor (no external behavior change)
- [ ] Documentation
- [ ] Tooling / CI

## Author checklist
- [x] `pnpm test` passes locally (326 tests, was 280 pre-merge)
- [x] `pnpm test:coverage` meets the thresholds (gate 94/77/98/95)
- [x] `pnpm check` is clean (lint + format, with `--error-on-warnings`)
- [x] `pnpm typecheck` passes
- [x] I have read the diff myself, line by line
- [x] I ran a Claude Code review on the diff and addressed its findings
- [x] Documentation is updated where needed (README split into local / remote quick starts, env var table, CLI reference, experimental banner)
- [x] If this PR adds an MCP tool: it is documented in `README.md` (n/a: no new tools, only a new transport)

## What is now possible

- Run a private remote MCP server for your Zendesk account. Point any modern MCP client (Claude Desktop, Claude Code, claude.ai web, ChatGPT, Gemini, Copilot, Cursor, VS Code, Windsurf, Zed, Perplexity, Mistral) at `https://your-server/mcp`. It discovers the OAuth flow against Zendesk automatically and runs every tool call with the end user's permissions.
- Stdio still works exactly as before. No regression for the laptop plus Claude Desktop scenario; the CI / headless API token escape hatch is preserved.
- Zero CORS config for the major web MCP clients. chatgpt.com, claude.ai, gemini.google.com, copilot.microsoft.com, perplexity.ai, chat.mistral.ai, grok.com, plus localhost on any port (MCP Inspector) are all allowed out of the box. Add your own with `--cors-origin`.
- Out-of-the-box discovery endpoints following the MCP 2025-06-18 spec: `/.well-known/oauth-protected-resource` (RFC 9728), `/.well-known/oauth-authorization-server` (RFC 8414, PKCE S256), plus `/healthz` for liveness checks.

## How you run it

| Use case | Command |
|---|---|
| Local stdio (laptop) | `npx -y @fruggr/zendesk-mcp-server acme` |
| Remote HTTP (server / container) | `zendesk-mcp-server acme --transport http --port 3000 --public-url https://mcp.example.com` |
| Azure App Service | `PUBLIC_URL="https://$WEBSITE_HOSTNAME" zendesk-mcp-server $ZENDESK_SUBDOMAIN --transport http --port $PORT` |
| Stdio + API token (CI / headless) | `ZENDESK_EMAIL=... ZENDESK_API_TOKEN=... zendesk-mcp-server acme` |

## What you still own as operator

The server provides the MCP transport, the OAuth discovery metadata and CORS for the major clients. The deployment side is yours:

- TLS termination (reverse proxy: Caddy, nginx, Cloudflare Tunnel, Azure App Service)
- Network exposure and firewall (the server binds `0.0.0.0` by default)
- Process supervision (systemd, Docker, Fly, your runner. None is shipped here)

## Notes for the reviewer (human or AI)

- **HTTP transport is shipped as experimental** (see banner at the top). The same notice is in the README. The outstanding validation is end-to-end OAuth from each supported MCP client against a real Zendesk tenant.
- **Per-proxy dispatch scoping** in `src/server.ts` (`buildProxyDispatch`) is a security fix vs the prior global `handlerMap`: previously `zendesk_tickets` with `operation: "get_article"` would have dispatched a help-center handler. Now each proxy carries its own scoped map. Covered by `tests/unit/server.test.ts` ("namespace proxy dispatch rejects operations outside its scoped tools").
- **HTTP 401 backstop**: when Zendesk rejects the bearer mid-call, the server surfaces it as an MCP tool error (not a 500), keeps the session alive, and does NOT try to invalidate any token store (there isn't one in HTTP). Locked by `tests/unit/transports/http.test.ts` ("surfaces a Zendesk 401 as an MCP tool error and keeps the session open").
- **CORS allowlist**: hardcoded constant (`DEFAULT_BROWSER_MCP_CLIENT_ORIGINS`) ordered by user base (ChatGPT first). The reflection path returns the matched constant string, not `req.headers.origin`, to satisfy CodeQL's flow analysis.
- **API token in HTTP is refused at boot**: `src/config.ts` throws if both `ZENDESK_EMAIL` and `ZENDESK_API_TOKEN` are set with `--transport http`. Tested.
- **No FastMCP / no AsyncLocalStorage**: an earlier draft used FastMCP; it was dropped after a security/maintenance review (~30 transitive deps, single maintainer, mcp-proxy patched `node:http` low enough to break MSW SSE). The current implementation is SDK-direct with a per-session `McpServer` whose bearer is captured in the tools' closure. CodeRabbit's auto-generated walkthrough still describes the FastMCP / ALS architecture because reviews are paused on this branch; it will refresh on the next `@coderabbitai review`.

## Verification

- [x] `pnpm check` / `pnpm typecheck` / `pnpm test`: 326 tests pass
- [x] `pnpm test:smoke`: both `stdio_transport_ready` and `http_transport_ready` markers observed
- [x] Manual: `curl /healthz`, `curl /.well-known/oauth-*`, and a real `claude mcp add` flow against a local server
- [x] CodeQL clean (the CORS credentials check is satisfied by the constant allowlist dataflow)
- [ ] End-to-end OAuth flow against a real Zendesk tenant from every supported MCP client. Outstanding; tracked by the experimental banner above.

## Follow-ups (not in this PR)

- [#41](https://github.com/fruggr/zendesk-mcp-server/issues/41) ready-to-deploy Docker image
- [#43](https://github.com/fruggr/zendesk-mcp-server/issues/43) per-session HTTP overrides via query params
- [#52](https://github.com/fruggr/zendesk-mcp-server/issues/52) end-to-end HTTP integration test once the MSW+SSE setup is in place

https://claude.ai/code/session_01KnSnQi68xNjBUqfUMfaZBT

### CODE REVIEW

**Comment by dlecan** [comment]
**Heads-up: #50 is landing on `main` — integration harness + coverage gate.**

When you rebase `main`, you inherit from #50:

1. **A transport-agnostic integration harness** (`tests/integration/`) built for this PR: implement `IntegrationHarness` in a new `http-harness.ts`, then `registerCoreScenarios(httpHarness)` in `http.test.ts` — the full MCP roundtrip suite runs over Streamable HTTP for free (see `stdio.test.ts`, 3 lines). Add HTTP-specific tests on top: OAuth discovery endpoints, `/healthz`, the `session-token` / `AsyncLocalStorage` bearer plumbing, and the API-token-refused-in-HTTP boot error.

2. **A coverage gate** in `vitest.config.ts` (statements 94 / branches 77 / functions 98 / lines 95). Only `index.ts`, `transports/stdio.ts`, `types.ts` are excluded — your HTTP code **is** measured, so without those tests CI fails on this PR. Cover the new code rather than lowering the thresholds.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: README.md [hunk 1]
```diff
@@ -2,20 +2,28 @@
...
+- **Two deployment shapes, same auth story** — Run it on your laptop as a stdio MCP server (Claude Desktop / Claude Code / VS Code) or deploy it as a private remote MCP server with one user, one Zendesk session per HTTP request.
 - **Context-friendly tool modes** — Expose 37 individual tools, 3 namespace proxies, or a single unified tool. Choose the mode that fits your LLM's context budget.
 - **Section-based article editing** — For large Help Center articles, read and rewrite one section at a time (parsed by h1/h2/h3 headings) instead of shuffling the full HTML body through the LLM. Reduces tokens by 10–100× on targeted edits.
 - **Read-only mode** — Restrict the server to read operations only, ideal for assistants that should never modify data.
-- **Zero runtime dependencies beyond the MCP SDK** — Built on `@modelcontextprotocol/sdk` and `zod`. No Express, no heavyweight frameworks.
+- **Lean stack** — Built on the official `@modelcontextprotocol/sdk` plus `zod`. The HTTP layer is a thin `node:http` wrapper around the SDK's `StreamableHTTPServerTransport` — no Express, no Hono, no framework on top of the framework.
```
Comment (dlecan) [inline README.md]: Issue: too verbose. To cut after zod

File: README.md [hunk 2]
```diff
@@ -123,27 +131,23 @@ zendesk-mcp-server acme --namespace tickets
...
-```bash
+# Or install globally
 npm install -g @fruggr/zendesk-mcp-server
 zendesk-mcp-server <your-subdomain>
-```
 
-Or clone and run locally:
-
-```bash
+# Or clone and run from source
```
Comment (dlecan) [inline README.md]: Nitpick: remove "developer" use to a dedicated section or remove it if redundant

File: README.md [hunk 3]
```diff
@@ -157,64 +161,145 @@ npx -y github:fruggr/zendesk-mcp-server <your-subdomain>
 npx -y github:fruggr/zendesk-mcp-server#my-feature-branch <your-subdomain>
 ```
 
-## Authentication
+### Zendesk OAuth setup
 
-The server supports two authentication methods:
+1. Go to **Admin Center → Apps and integrations → APIs → OAuth Clients**
```
Comment (dlecan) [inline README.md:172]: Suggestion: explain a user with admin rights must do that

File: README.md [hunk 4]
```diff
@@ -157,64 +161,145 @@ npx -y github:fruggr/zendesk-mcp-server <your-subdomain>
 npx -y github:fruggr/zendesk-mcp-server#my-feature-branch <your-subdomain>
 ```
 
-## Authentication
+### Zendesk OAuth setup
 
-The server supports two authentication methods:
```
Comment (dlecan) [inline README.md:182]: Question : that was interesting information 
Does it exist anywhere else?

File: README.md [hunk 5]
```diff
@@ -157,64 +161,145 @@ npx -y github:fruggr/zendesk-mcp-server <your-subdomain>
...
 
-For headless/CI environments or quick testing.
+Deploy a private MCP server for **one** Zendesk account. Every MCP client connecting to the server presents its **own** user's OAuth bearer in `Authorization:` — the server never sees a shared admin key.
 
-**Zendesk setup:**
+### Zendesk OAuth setup
 
-1. Go to **Admin Center > Apps and integrations > APIs > Zendesk API**
-2. Enable **Token Access**, create a token
+1. Go to **Admin Center → Apps and integrations → APIs → OAuth Clients**
```
Comment (dlecan) [inline README.md]: Question : duplicated, isn't it?

File: README.md [hunk 6]
```diff
@@ -330,6 +461,27 @@ pnpm check
 pnpm test
 ```
 
+## Appendix: API token authentication (stdio only)
```
Comment (dlecan) [inline README.md]: Suggestion: move this to an external doc

File: AGENTS.md [hunk 7]
```diff
@@ -1,30 +1,28 @@
...
 ```bash
 pnpm install
 ```
 
+Toolchain (Node 24 + pnpm 11) is the dev floor; the published package still runs on Node 20+ (a CI job exercises the packed tarball on Node 20).
+
 ## Architecture
 
 ```
 src/
```
Comment (dlecan) [inline AGENTS.md]: Nitpick: in another pr, this structure has been summarized, this the global structure (1st folder level ?)
Reveal structure will be crawled by this agent anyway

File: AGENTS.md [hunk 8]
```diff
@@ -51,104 +52,66 @@ Tools are registered at startup based on `--mode`:
...
+`--namespace` and `--read-only` are applied by `filterTools` (`src/routing/registry.ts`) *before* the mode switch in `src/server.ts`. They therefore narrow every mode, including the default `namespace` mode. `--tool <name>` is also filtered here but additionally forces `mode: 'all'` in `src/config.ts`.
 
 ### Token passing
 
-In API token mode, a static Basic auth header is built from `ZENDESK_EMAIL` + `ZENDESK_API_TOKEN`. In OAuth mode, the token is obtained on-demand via browser PKCE flow and cached in memory by `token-store.ts`. Both modes pass a `getToken` function to tool handlers.
+`createMcpServer(config, getToken)` injects a single `getToken: () => string | Promise<string>` closure into every tool handler. Where the closure pulls the token from depends on the transport:
```
Comment (dlecan) [inline AGENTS.md]: Nitpick: summarize the section

File: AGENTS.md [hunk 9]
```diff
@@ -37,12 +35,15 @@ src/
...
+│   ├── stdio.ts          # SDK StdioServerTransport wrapper
+│   └── http.ts           # node:http server: /mcp (SDK StreamableHTTPServerTransport), /.well-known/oauth-*, /healthz, per-session McpServer
 └── utils/
     ├── formatting.ts     # Markdown formatters per entity type
     └── pagination.ts     # Cursor-based pagination helpers
 ```
 
+Transports use the official `@modelcontextprotocol/sdk` directly. Stdio is `StdioServerTransport`; HTTP is a thin `node:http` server wrapping `StreamableHTTPServerTransport` plus the OAuth discovery endpoints (RFC 9728 / RFC 8414) advertising Zendesk as the upstream authorization server per MCP Specification 2025-06-18. No third-party MCP framework on top.
+
 ### Tool modes (pattern Azure MCP Server)
```
Comment (dlecan) [inline AGENTS.md]: Nitpick : summarize the section

File: AGENTS.md [hunk 10]
```diff
@@ -51,104 +52,66 @@ Tools are registered at startup based on `--mode`:
...
-   pnpm dev -- <your-subdomain> --mode all
-   ```
-4. On the first tool call, a browser window opens for authentication.
+# Stdio + API token (only when a browser is not available)
+ZENDESK_EMAIL=you@example.com ZENDESK_API_TOKEN=xxx \
+  pnpm dev -- <your-subdomain> --mode all
+```
 
-### Option B: API token
+Zendesk-side OAuth client setup (admin center, redirect URLs, tokens) is documented in [`README.md`](README.md#quick-start-local-stdio).
```
Comment (dlecan) [inline AGENTS.md]: Praise: nice, make a reference to an external doc, here the readme

File: AGENTS.md [hunk 11]
```diff
@@ -51,104 +52,66 @@ Tools are registered at startup based on `--mode`:
...
 
-1. Go to **Admin Center → Apps and integrations → APIs → Zendesk API**
-2. Enable **Token Access** in Settings tab
-3. Create an API token
-4. Run:
-   ```bash
-   ZENDESK_EMAIL=you@example.com ZENDESK_API_TOKEN=xxx \
-     pnpm dev -- <your-subdomain> --mode all
-   ```
+## Smoke-testing a tool manually
```
Comment (dlecan) [inline AGENTS.md]: Nitpick: not necessary anymore, smoke tests are automatic

File: AGENTS.md [hunk 12]
```diff
@@ -182,51 +146,29 @@ replies). Direct chat with the user follows their language.
...
-   once dev is done and the local gate is green — never leave it as a draft.
-
-The maintainer's review starts from the assumption that everything above
-has already been done.
+1. **Re-read your own diff in full.** No skimming. If a hunk no longer makes sense out of the context where you wrote it, rewrite it.
+2. **Justify each change.** For every non-trivial hunk, you should be able to answer: why is this change here, what would break without it, and is it the smallest version of the fix.
+3. **Look for what you didn't write.** Missing zod validation on an input, missing test for an edge case, missing README/AGENTS update on a renamed tool, missing error path. Reviewers find these — find them first.
+4. **Self-review prompt.** Run a Claude Code pass on the diff against `main` using the prompt in [`CONTRIBUTING.md`](CONTRIBUTING.md#author-side-ai-review) "Author-side AI review". Address findings or document why you're skipping them in the PR description.
+5. **Run the full local gate**: `pnpm check`, `pnpm typecheck`, `pnpm test`, `pnpm test:coverage`, `pnpm build`, `pnpm test:smoke`. A green CI on a non-green local run means a flaky check, not a free pass.
```
Comment (dlecan) [inline AGENTS.md]: Question : explained twice in the file (commands are listed at the beginning of the file)
Useful?
This one seems better

File: AGENTS.md [hunk 13]
```diff
@@ -182,51 +146,29 @@ replies). Direct chat with the user follows their language.
...
 
-Any change to the tool surface requires a README sync in the same PR. The
-README is what external users rely on — it drifts fast if ignored.
+Any change to the tool surface requires a README sync in the same PR. The README is what external users rely on — it drifts fast if ignored.
 
 When you add, remove, rename, or meaningfully re-describe a tool, update:
 
-- **`README.md`** — the matching row in the `Tickets` / `Help Center` / `Users & Organizations` / `Search` table, the `(N tools)` count in the `<summary>`, and the global tool count (currently **36**) wherever it appears ("Expose 36 individual tools", mode table, single-mode tip, CLI example).
-- **`AGENTS.md`** — the per-file tool counts in the Architecture section (`tickets.ts # 9 ticket tools`, `help-center.ts # 21 Help Center tools`, `users.ts # 5 user/organization tools`) and the `all` mode line in "Tool modes".
+- **`README.md`** — the matching row in the `Tickets` / `Help Center` / `Users & Organizations` / `Search` table, the `(N tools)` count in the `<summary>`, and the global tool count (currently **37**) wherever it appears.
```
Comment (dlecan) [inline AGENTS.md]: Question : you know this file? Is it necessary?

File: AGENTS.md [hunk 14]
```diff
@@ -182,51 +146,29 @@ replies). Direct chat with the user follows their language.
...
+- **`AGENTS.md`** — the per-file tool counts in the Architecture section above.
 - Namespace counts in `tests/unit/routing/registry.test.ts` if you touch the `help_center`, `tickets`, or `users` namespace.
```
Comment (dlecan) [inline AGENTS.md]: Nitpick: not necessary

File: AGENTS.md [hunk 15]
```diff
@@ -182,51 +146,29 @@ replies). Direct chat with the user follows their language.
...
 - The `createHelpCenterTools` length assertion in `tests/unit/tools/help-center.test.ts` (and equivalents for other namespaces).
```
Comment (dlecan) [inline AGENTS.md]: Nitpick: not necessary

File: README.md [hunk 16]
```diff
@@ -143,58 +150,35 @@ zendesk-mcp-server acme --namespace tickets
...
-2. Create a public client:
+1. Go to **Admin Center → Apps and integrations → APIs → OAuth Clients**
+2. Create a **public** client:
    - **Identifier**: `<your-subdomain>_zendesk` (or set `ZENDESK_OAUTH_CLIENT_ID`)
    - **Redirect URL**: `http://localhost:27439/callback` (change the port to match
      `ZENDESK_OAUTH_CALLBACK_PORT` / `--callback-port` if you override it; Zendesk
      accepts several redirect URLs, one per line)
 
-**Run:**
+On the first tool call, a browser window opens for OAuth login. The token is cached in memory for the session.
```
Comment (dlecan) [inline README.md]: Question : "in memory" is that true?

File: README.md [hunk 17]
```diff
@@ -220,82 +204,208 @@ browser sign-in.
...
+`--public-url` (or `PUBLIC_URL=…`) is the URL **clients use to reach you**. It's what gets advertised in the OAuth discovery metadata as the canonical resource identifier (RFC 8707). When the server is behind a TLS reverse proxy — Azure App Service, Heroku, Fly.io, Cloudflare Tunnel, nginx, Caddy… — the bind host and the public URL differ, and spec-compliant MCP clients will refuse the connection if the metadata advertises the wrong resource. Without it the server boots in a degraded mode and prints a warning.
+
+| Platform | Recommended setup |
+|---|---|
+| **Azure App Service** | Startup command: `PUBLIC_URL="https://$WEBSITE_HOSTNAME" zendesk-mcp-server $ZENDESK_SUBDOMAIN --transport http --port $PORT` |
+| **Heroku / Fly / Cloud Run** | `PUBLIC_URL=https://<your-app>.<provider>.app` in the env / config |
+| **Caddy / nginx / Traefik in front of a VM** | `PUBLIC_URL=https://mcp.example.com` |
+| **Local dev (no proxy)** | `--host 127.0.0.1 --port 3000` — the resource URL is derived automatically (the wildcard `0.0.0.0` is what triggers the warning) |
+
+### Sessions & request limits
```
Comment (dlecan) [inline README.md]: Question: that's internal details. Should they be documented in the primary readme ?
